developer_guides/testbench/debug_in_testbench.rst

```diff
@@ -0,0 +1,214 @@
+
+Debug component in Testbench
+############################
+
+GDB and DDD
+***********
+
+Code debugging with debugger is an efficient way of finding issues in
+components. The code may crash or operate incorrectly. Also the SOF
+data structures can be understood better while seeing them in action.
+
+In testbench environment a severe memory access mistake typically
+results to a segmentation fault where the operating system traps the
+application when it performs illegal memory access to RAM it has not
+allocated. The debugger shows the stack trace of calls if this happens
+for quick spotting of offending code.
+
+A stable but incorrectly working component can be examined with
+breakpoints and visualization of data structures.
+
+To initiate debugging the output from previous command to run IIR EQ
+is studied. The information for debugging is shown in the beginning.
+
+.. code-block:: bash
+
+   ./eqiir_run.sh 16 16 48000 audio_in.raw audio_out.raw
+
+::
+
+   Command:         ../../testbench/build_testbench/install/bin/testbench
+   Argument:        -d -r 48000 -R 48000 -i audio_in.raw -o audio_out.raw -t ../../build_tools/test/topology/test-playback-ssp5-mclk-0-I2S-eq-iir-s16le-s16le-48k-24576k-codec.tplg -b S16_LE 
+   LD_LIBRARY_PATH: ../../testbench/build_testbench/sof_ep/install/lib:../../testbench/build_testbench/sof_parser/install/lib
+
+In the above output the command shows the path to installed testbench
+binary. The arguments specify e.g. the input and output sample rate,
+input and output RAW data files, topology to use for testing and
+sample format. The command line options are described when invoking
+the binary with switch "-h". But for the binary to work correctly the
+dynamic libraries path need to be instructed for the operating
+system. It is done by setting the environment variable
+LD_LIBRARY_PATH to above shown value.
+
+.. code-block:: bash
+
+   export LD_LIBRARY_PATH=../../testbench/build_testbench/sof_ep/install/lib:../../testbench/build_testbench/sof_parser/install/lib
+   ../../testbench/build_testbench/install/bin/testbench -h
+
+If the help text appeared the testbench binary start directly from
+command line worked. Next, the testbench can be started in Data
+Display Debugger (DDD) application. DDD is a graphical front-end for
+GNU Debugger (GDB). DDD and the dependencies such as GDB needs to be
+installed if missing from development computer.
+
+.. code-block:: bash
+
+   sudo apt install ddd
+
+The debugging is started to previously used shell with LD_LIBRARY_PATH set.
+
+.. code-block:: bash
+
+   ddd ../../testbench/build_testbench/install/bin/testbench
+
+This opens the debugger window. From there find the code line just
+after topology parsing (currently 295) by scrolling the code window
+with mouse and place a break point there with right mouse button (a
+red stop sign). If there are issues that happen at topology parsing
+or ẃithin component in instantiating in new() place the breakpoint to
```

Comment:
there's some strange accent in the w here...

Reply:
Oops!